hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala

```diff
@@ -257,6 +261,147 @@ object EmitStream {
+            val maxSize = mb.genFieldThisRef[Int]("stream_buff_agg_max_size")
+            val nodeArray = mb.genFieldThisRef[Array[Long]]("stream_buff_agg_element_array")
+            val idx = mb.genFieldThisRef[Int]("stream_buff_agg_idx")
+            val returnStreamType= x.typ.asInstanceOf[TStream]
+            val returnElemType = returnStreamType.elementType
+            val tupleFieldTypes = aggSignatures.map(_ => TBinary)
+            val tupleFields = (0 to tupleFieldTypes.length).zip(tupleFieldTypes).map { case (fieldIdx, fieldType) => TupleField(fieldIdx, fieldType) }
+            val serializedAggSType = SStackStruct(TTuple(tupleFields), tupleFieldTypes.map(_ => EmitType(SBinaryPointer(PCanonicalBinary()), true)).toIndexedSeq)
+            val keyAndAggFields = newKeyVType.canonicalPType.asInstanceOf[PCanonicalStruct].sType.fieldEmitTypes :+ EmitType(serializedAggSType, true) // TODO is this required?
+            val returnElemSType = SStackStruct(returnElemType.asInstanceOf[TBaseStruct], keyAndAggFields)
+            val newStreamElem = mb.newEmitField("stream_buff_agg_new_stream_elem", EmitType(returnElemSType, true))
+            val numElemInArray = mb.genFieldThisRef[Int]("stream_buff_agg_num_elem_in_size")
+            val childStreamEnded = mb.genFieldThisRef[Boolean]("stream_buff_agg_child_stream_ended")
+            val produceElementMode = mb.genFieldThisRef[Boolean]("stream_buff_agg_child_produce_elt_mode")
+            val producer: StreamProducer = new StreamProducer {
+              override val length: Option[EmitCodeBuilder => Code[Int]] = None
+
+              override def initialize(cb: EmitCodeBuilder): Unit = {
+                if (childProducer.requiresMemoryManagementPerElement)
+                  cb.assign(childProducer.elementRegion, Region.stagedCreate(Region.REGULAR, outerRegion.getPool()))
+                else
+                  cb.assign(childProducer.elementRegion, region)
+
+                childProducer.initialize(cb)
+                cb.assign(childStreamEnded, false)
+                cb.assign(produceElementMode, false)
+                cb.assign(idx, 0)
+                cb.assign(maxSize, 8)
+                cb.assign(nodeArray, Code.newArray[Long](maxSize))
+                cb.assign(numElemInArray, 0)
+                dictState.createState(cb)
+              }
+
+              override val elementRegion: Settable[Region] = region
+
+              override val requiresMemoryManagementPerElement: Boolean = false
+
+
+              override val LproduceElement: CodeLabel = mb.defineAndImplementLabel { cb =>
+                val elementProduceLabel = CodeLabel()
+                val getElemLabel = CodeLabel()
+                val startLabel = CodeLabel()
+
+                cb.define(startLabel)
+                cb.ifx(produceElementMode, {
+                  cb.goto(elementProduceLabel)
+                })
+
+                dictState.newState(cb)
```

Comment:
can you add a comment before this line that it GCs the old aggregator states if we're moving onto a new group?